File: app/src/main/java/com/fanchen/imovie/base/BaseTabActivity.java
package com.fanchen.imovie.base;

import android.os.Bundle;
import android.support.design.widget.TabLayout;
import android.support.v4.app.FragmentManager;
import android.support.v4.view.PagerAdapter;
import android.support.v4.view.ViewPager;
import android.view.LayoutInflater;
import android.view.View;

import com.fanchen.imovie.R;
import com.fanchen.imovie.view.NoScrollViewPager;

import butterknife.ButterKnife;
import butterknife.InjectView;

/**
 * Created by fanchen on 2017/7/24.
 */
public abstract class BaseTabActivity extends BaseToolbarActivity {

    @InjectView(R.id.sliding_tabs)
    protected TabLayout mSlidingTab;
    @InjectView(R.id.view_pager)
    protected NoScrollViewPager mViewPager;

    @Override
    protected int getLayout() {
        return R.layout.activity_tab_pager;
    }

    @Override
    protected void initActivity(Bundle savedState, LayoutInflater inflater) {
        super.initActivity(savedState, inflater);
        PagerAdapter adapter = getAdapter(getSupportFragmentManager());
        mViewPager.setAdapter(adapter);
        mViewPager.setOffscreenPageLimit(adapter.getCount());
        mSlidingTab.setupWithViewPager(mViewPager);
        mSlidingTab.setTabMode(getTabMode());
    }

    protected int getTabMode(){
        return TabLayout.MODE_FIXED;
    }

    /**
     * @return
     */
    public ViewPager getViewPager() {
        return mViewPager;
    }

    /**
     * @return
     */
    public TabLayout getSlidingTab() {
        return mSlidingTab;
    }

    /**
     * @return
     */
    protected abstract PagerAdapter getAdapter(FragmentManager fm);

}

File: app/src/main/java/com/fanchen/imovie/base/BaseTabActivity.kt
package com.fanchen.imovie.base

import android.os.Bundle
import android.support.design.widget.TabLayout
import android.support.v4.app.FragmentManager
import android.support.v4.view.PagerAdapter
import android.support.v4.view.ViewPager
import android.view.LayoutInflater
import com.fanchen.imovie.R
import kotlinx.android.synthetic.main.activity_tab_pager.*

abstract class BaseTabActivity : BaseToolbarActivity() {
    override fun getLayout(): Int {
        return R.layout.activity_tab_pager
    }

    var mViewPager:ViewPager? = null

    override fun initActivity(savedState: Bundle?, inflater: LayoutInflater) {
        super.initActivity(savedState, inflater)
        val adapter = getAdapter(supportFragmentManager)
        if (!checkTabViewNull()) {
            view_pager.adapter = adapter
            view_pager.offscreenPageLimit = adapter.count
            sliding_tabs.setupWithViewPager(view_pager)
            sliding_tabs.tabMode = getTabMode(adapter)
            mViewPager = view_pager
        }
    }

    protected open fun getTabMode(adapter: PagerAdapter): Int {
        return if (adapter.count < 5) TabLayout.MODE_FIXED else TabLayout.MODE_SCROLLABLE
    }

    /**
     * @return
     */
    protected abstract fun getAdapter(fm: FragmentManager): PagerAdapter

    private fun checkTabViewNull(): Boolean {
        return sliding_tabs == null || view_pager == null
    }
}
